src/EnergyPlus/UserDefinedComponents.cc

```diff
@@ -1160,7 +1160,7 @@ namespace UserDefinedComponents {
                                                                 cAlphaArgs(1),
                                                                 DataLoopNode::NodeFluidType::Water,
                                                                 DataLoopNode::ConnectionType::Inlet,
-                                                                NodeInputManager::CompFluidStream::Secondary,
+                                                                NodeInputManager::CompFluidStream::Tertiary,
```

Comment:
For Coil:UserDefined, I understand the change to the air-side fluid stream type above but for Plant Connection Inlet Node, A9, why is this changed from Secondary to Tertiary? I only see the word Tertiary used in the EvaporativeCoolers code so am wondering what this node fluid stream connection type represents. 

            EvapCond(EvapCoolNum).TertiaryInletNode = GetOnlySingleNode(state,
                                                                        state.dataIPShortCut->cAlphaArgs(12),
                                                                        ErrorsFound,
                                                                        cCurrentModuleObject,
                                                                        state.dataIPShortCut->cAlphaArgs(1),
                                                                        DataLoopNode::NodeFluidType::Air,
                                                                        DataLoopNode::NodeConnectionType::Inlet,
                                                                        3,
                                                                        ObjectIsNotParent);
 

The water coils use Secondary and I would think this component would use the same?

        state.dataWaterCoils->WaterCoil(CoilNum).WaterInletNodeNum = GetOnlySingleNode(state,
                                                                                       AlphArray(3),
                                                                                       ErrorsFound,
                                                                                       CurrentModuleObject,
                                                                                       AlphArray(1),
                                                                                       DataLoopNode::NodeFluidType::Water,
                                                                                       DataLoopNode::NodeConnectionType::Inlet,
                                                                                       2,
                                                                                       ObjectIsNotParent);
        state.dataWaterCoils->WaterCoil(CoilNum).WaterOutletNodeNum = GetOnlySingleNode(state,
                                                                                        AlphArray(4),
                                                                                        ErrorsFound,
                                                                                        CurrentModuleObject,
                                                                                        AlphArray(1),
                                                                                        DataLoopNode::NodeFluidType::Water,
                                                                                        DataLoopNode::NodeConnectionType::Outlet,
                                                                                        2,
                                                                                        ObjectIsNotParent);

Reply:
My understanding is that these connection types have "local" meanings to each particular object. It probably just trying to identify each object has how many fluid streams through it. So for the water coil only primary and secondary are needed since there is only one air fluid stream and one water stream. 

Across different objects, my understanding (maybe incorrect?) is that these stream tags are not universally aligned--e.g. one object's secondary fluid stream (or even primary stream) does not mean it has to be on the same stream on another object's secondary fluid stream. 

On thing for sure is that there are two inlets and two outlet streams on the same fluid stream for the same object, the Node connection Check 10 would give a fatal error--that is what is for. So for the Coil:UserDefined, if there there are primary and secondary air stream (secondary air stream input is optional for this object), and if the water stream is assigned to secondary again, the node connection check 10 would not pass.

Reply:
Another example of Tietiary (or even Quaterary) fluid stream is AirTerminal:SingleDuct:UserDefined. By looking at the following code, it would be inferred that there could be automatically have Tietiary water side flow connections, or Quaterary connections (this is also one of the reasons why a "Quaterary" fluid stream type enum is added in this PR):
``` 
                int NumPlantConnections = std::floor(rNumericArgs(1));
                state.dataUserDefinedComponents->UserAirTerminal(CompLoop).NumPlantConnections = NumPlantConnections;
                if ((NumPlantConnections >= 1) && (NumPlantConnections <= 2)) {
                    state.dataUserDefinedComponents->UserAirTerminal(CompLoop).Loop.allocate(NumPlantConnections);
                    for (int ConnectionLoop = 1; ConnectionLoop <= NumPlantConnections; ++ConnectionLoop) {
                        int aArgCount = (ConnectionLoop - 1) * 2 + 8;
                        state.dataUserDefinedComponents->UserAirTerminal(CompLoop).Loop(ConnectionLoop).InletNodeNum =
                            NodeInputManager::GetOnlySingleNode(state,
                                                                cAlphaArgs(aArgCount),
                                                                ErrorsFound,
                                                                DataLoopNode::ConnectionObjectType::AirTerminalSingleDuctUserDefined,
                                                                cAlphaArgs(1),
                                                                DataLoopNode::NodeFluidType::Water,
                                                                DataLoopNode::ConnectionType::Inlet,
                                                                static_cast<NodeInputManager::CompFluidStream>(ConnectionLoop + 2),
                                                                DataLoopNode::ObjectIsNotParent,
                                                                cAlphaFieldNames(aArgCount));
                        state.dataUserDefinedComponents->UserAirTerminal(CompLoop).Loop(ConnectionLoop).OutletNodeNum =
                            NodeInputManager::GetOnlySingleNode(state,
                                                                cAlphaArgs(aArgCount + 1),
                                                                ErrorsFound,
                                                                DataLoopNode::ConnectionObjectType::AirTerminalSingleDuctUserDefined,
                                                                cAlphaArgs(1),
                                                                DataLoopNode::NodeFluidType::Water,
                                                                DataLoopNode::ConnectionType::Outlet,
                                                                static_cast<NodeInputManager::CompFluidStream>(ConnectionLoop + 2),
                                                                DataLoopNode::ObjectIsNotParent,
                                                                cAlphaFieldNames(aArgCount + 1));
```

Reply:
Do these tags affects how the overall air loops or plants are simulated? Maybe--but it seems that if within each object handles the nodes properly (hopefully in simulation they do not only relay on this tag type to identify fluid stream across elements), it should be fine. I checked the Coil:UserDefined code this does not seem to be a problem. Maybe some comments are welcome about if this could be an issue to other objects?

Reply:
I think the FluidStream is used to help draw the HVACDiagram. I don't think it is used anywhere in E+. It would be better if @mjwitte or @JasonGlazer commented on how this data is used.

Reply:
The FluidStream is used in EnergyPlus for connection checking for each object, and it is used in HVACDiagram, but it's just a way to distinguish different fluid streams within a given object to check if inlets and outlets make sense. HVACDiagram doesn't care about the actual integer values for this, it just cares that inlets a and b go together on the same stream, and so on. There can be a gap in the numbering on a given object. Before this was `enum`ized it was just an integer and there were hard numbers in the `getOnlySingleNode` calls, or, in this case there were loop indices which is how this got broken during the `enum` sweep.

Reply:
Ok, I see the node connection check [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/BranchNodeConnections.cc#L122-L129). This code is looking for another node already registered that matches the same FluidStream for the same object. The 3rd check is for ConnectionType (inlet or outlet) and since Coil:UserDefined can have 3 sets of node pairs, this check gets tripped up. It seems there should not be an issue between air and water nodes, but because this check does not look for NodeFluidType, then the node connection check fails for this object if 2 air streams and a water stream are used (so you can't use secondary here for FluidStream). Since the integer value does not matter then it seems this is OK.

Reply:
I'm still suspicous of this change because the first check in this block of code is for NodeNumber. The air and water node numbers should not match. I'll leave it there.

Reply:
The "Check 10" below will check inlets and outlets of an object and accumulate counts into fluidstream "bins". There the air and the water streams are not distinguished other than "fluidstream" tag numbers: 
https://github.com/NREL/EnergyPlus/blob/44de4f7d8067ad59f1bb2ed1a4c6e83740eb3c6b/src/EnergyPlus/BranchNodeConnections.cc#L698-L722 

If >1 inlets and >1 outlets are detected at the same time in the same fluidstream bin, the error will be thrown. So air secondary + water secondary should also trigger this error. The original code triggers the error slightly differently, by doing "air primary + air primary".

Reply:
I'm glad this conversation wrapped up nicely :smile: